framework/py/flwr/supercore/object_store/in_memory_object_store.py

```diff
@@ -69,41 +112,120 @@ def put(self, object_id: str, object_content: bytes) -> None:
             validate_object_content(content=object_content)
 
-        # Return if object is already present in the store
-        if self.store[object_id] != b"":
-            return
+        with self.lock_store:
+            # Only allow adding the object if it has been preregistered
+            if object_id not in self.store:
+                raise NoObjectInStoreError(
+                    f"Object with ID '{object_id}' was not pre-registered."
+                )
 
-        self.store[object_id] = object_content
+            # Return if object is already present in the store
+            if self.store[object_id].is_available:
+                return
+
+            # Update the object entry in the store
+            self.store[object_id].is_available = True
+            self.store[object_id].content = object_content
 
     def set_message_descendant_ids(
         self, msg_object_id: str, descendant_ids: list[str]
     ) -> None:
         """Store the mapping from a ``Message`` object ID to the object IDs of its
         descendants."""
-        self.msg_children_objects_mapping[msg_object_id] = descendant_ids
+        with self.lock_msg_mapping:
+            self.msg_descendant_objects_mapping[msg_object_id] = descendant_ids
 
     def get_message_descendant_ids(self, msg_object_id: str) -> list[str]:
         """Retrieve the object IDs of all descendants of a given Message."""
-        if msg_object_id not in self.msg_children_objects_mapping:
-            raise NoObjectInStoreError(
-                f"No message registered in Object Store with ID '{msg_object_id}'. "
-                "Mapping to descendants could not be found."
-            )
-        return self.msg_children_objects_mapping[msg_object_id]
+        with self.lock_msg_mapping:
+            if msg_object_id not in self.msg_descendant_objects_mapping:
+                raise NoObjectInStoreError(
+                    f"No message registered in Object Store with ID '{msg_object_id}'. "
+                    "Mapping to descendants could not be found."
+                )
+            return self.msg_descendant_objects_mapping[msg_object_id]
+
+    def delete_message_descendant_ids(self, msg_object_id: str) -> None:
+        """Delete the mapping from a ``Message`` object ID to its descendants."""
+        with self.lock_msg_mapping:
+            self.msg_descendant_objects_mapping.pop(msg_object_id, None)
 
     def get(self, object_id: str) -> Optional[bytes]:
         """Get an object from the store."""
-        return self.store.get(object_id)
+        with self.lock_store:
+            # Check if the object ID is pre-registered
+            if object_id not in self.store:
+                return None
+
+            # Check if the object is available (i.e., has been put into the store)
+            object_entry = self.store[object_id]
+            if not object_entry.is_available:
+                return b""
+
+            return object_entry.content
```

Comment:
```suggestion
            return object_entry.content
```

We are preregistering it with `b""` so we can return `.content` directly